AMM Application Contracts Port (#230)
FuelLabs/sway-applications: AMM/SPECIFICATION.md

    3. If the amount of liquidity pool assets are sufficient
    4. Requires the minimum amounts of both assets to receive after burning
    5. Requires a deadline (block height limit)

##### `withdraw()`

1. Allows withdrawing previously deposited assets without using them to add liquidity 
    1. If the asset pair of the pool is set
    2. If the deposited amount of the asset is sufficient
    3. Requires an amount of either asset to withdraw

##### `swap_with_exact_input()`

1. Allows selling an exact amount of an asset for the other asset 
    1. If the asset pair of the pool is set
    2. If the deadline has not passed
    3. If the asset reserve of asset to buy in the contract is sufficient
    4. If the bought amount is not less than the optional minimum output amount
    5. Requires an exact amount of either asset to sell
    6. Requires an optional minimum desired output amount
        > **NOTE** This is a safety mechanism against excessive slippage. The [`preview_swap_with_exact_input()`](#preview_swap_with_exact_input) function can be used to calculate a reasonable minimum output amount.
    7. Requires a deadline (block height limit)

> ##### `swap_with_exact_output()`

Reviewer: ```suggestion
##### `swap_with_exact_output()`

```

            - [`swap_with_exact_input()`](#swap_with_exact_input)
            - [`swap_with_exact_output()`](#swap_with_exact_output)
        - [Exchange Previews](#exchange-previews)
            - [`preview_add_liquidity()`](#preview_add_liquidity)
            - [`preview_swap_with_exact_input()`](#preview_swap_with_exact_input)
            - [`preview_swap_with_exact_output()`](#preview_swap_with_exact_output)
        - [Exchange State Checks](#exchange-state-checks)
            - [`balance()`](#balance)
            - [`pool_info()`](#pool_info)
- [Sequence Diagram](#sequence-diagram)

# Overview

This document provides an overview of the application.

It outlines the use cases, i.e. desirable functionality, in addition to requirements for the smart contracts.

# Use Cases

This section contains general information about the functionality of the application and thus does not touch upon any technical aspects.

If you are interested in a functional overview then this is the section for you.

## Actions that users are able to perform

This sub-section details what a user is able to do e.g. click a button and "x, y, z" happens.

### AMM Core Functionality

#### `initialize()`

> 1. Allows specifying the legitimate exchange contract implementation that the AMM will operate with Requires

Reviewer: 1. `Allows specifying`?
2. What is an illegitimate contract and how does the user know that information?
3. `that the AMM will operate with Requires`?

    1. The bytecode root of the desired exchange contract implementation

#### `add_pool()`

> 1. Allows an asset pair to be registered to the AMM 

Reviewer: ```suggestion
1. Creates a new liquidity pool with the specified asset pair
```

Author: The liquidity pool is created in the exchange contract, the AMM contract simply stores the mappings.

>     1. If the AMM is initialized Requires

Reviewer: ?

    2. Requires
        1. The identifiers of the two assets
        2. The exchange contract identifier that is also the identifier of the liquidity pool asset for the given pair if the exchange contract
            1. Is legitimate, i.e., the bytecode root matches

>             2. Defines the pool for the provided asset pair

Reviewer: These should be reworded to be separate points rather than a `requires` "tag" somewhere which the user needs to look for and figure out some sentence substitution. Each bullet point should be self contained and explicit

### AMM State Checks

#### `pool()`

> 1. Returns the registered exchange contract identifier for an asset pair Requires

Reviewer: Registered? Why would it return some "unregistered" id? I won't repeat the `requires` comment each time if there are more instances below

    1. The identifiers of the two assets

### Exchange Core Functionality

#### `constructor()`

> 1. Allows specifying the asset pair for which the pool will be managed through the exchange contract 

Reviewer: for which the pool will be managed?

    1. If the asset pair for the exchange contract has not already been set 
    2. Requires two different asset identifiers

#### `deposit()`

> 1. Allows depositing an asset to the contract (to later be used for adding liquidity or be withdrawn) 

Reviewer: No need for the "well maybe this thing in the future". That implication should only be stated in a section which explains the sequence of actions performed by the user rather than on a page which strictly documents the functionality of each function.

"Deposits an asset into the contract" although this still isn't great because despite what I've just said it is missing context. Perhaps your sentence is more suited at the end of this one though

Author: Changed "Allows depositing an asset to the contract" to "Deposits an asset into the contract". Can you be more specific with what your suggestion is with "Perhaps your sentence is more suited at the end of this one though"?

    1. If the asset pair of the pool is set 
    2. Requires any amount of either asset in the pair

#### `add_liquidity()`

> 1. Allows minting liquidity pool assets by using up the deposits when possible 

Reviewer: No need to mention minting because the user is not allowed to mint, if anything it automatically happens as part of a process, no?

`deposits`? Like, a counter for how many times a user has deposited?

`when possible`? uhm, when is it not possible? I assume when the requirements are not met in which case that is not needed and instils uncertainty. The docs should be statements to tell the reader what happens as a matter of fact and if there are exceptions then those requirements can also be listed as facts below

Author: Rephrased to clarify that "when possible" means one asset will be used up but the other one may not be used up due to the ratio of the two assets.

    1. If
        1. The asset pair of the pool is set
        2. The deadline has not passed

>         3. Deposit amounts of both assets are sufficient

Reviewer: What does `sufficient` mean here? Can you be more specific?

        4. The resulting liquidity amount is more than the minimum liquidity for a pool to exist. 
            > **NOTE** This lower limit for liquidity can be modified as desired per asset pair and exists to protect against excessive slippage.
    2. Requires 
        1. The desired liquidity amount

>         2. A deadline, i.e., block height limit

Reviewer: ```suggestion
        2. A deadline (block height limit)
```

This is probably fine

#### `remove_liquidity()`

1. Allows burning liquidity pool assets Requires 
    1. If
        1. The asset pair of the pool is set
        2. The deadline has not passed
        3. The amount of liquidity pool assets are sufficient
    2. Requires
        1. The minimum amounts of both assets to receive after burning
        2. A deadline, i.e., block height limit

#### `withdraw()`

1. Allows withdrawing previously deposited assets without using them to add liquidity 
    1. If
        1. The asset pair of the pool is set
        2. The deposited amount of the asset is sufficient
    2. Requires an amount of either asset to withdraw

#### `swap_with_exact_input()`

> 1. Allows selling an exact amount of an asset for the other asset 

Reviewer: Without looking at the code it's unclear which / what asset has some exact amount and what this `other` asset is. Can you be more specific i.e. if there are two assets that are provided as arguments then state the amount of asset A and that it gets swapped for the specified asset B etc.

Author: The input and output assets can be either one of asset A or asset B. This is specified with the function call.

    1. If 
        1. The asset pair of the pool is set
        2. The deadline has not passed
        3. The asset reserve of asset to buy in the contract is sufficient

>         4. The bought amount is not less than the optional minimum desired output amount

Reviewer: optimal... minimum... desired... output amount. That's a hell of a loaded sentence. What do all of these terms mean? Could you simplify / rephrase so that the intent behind each word is clear?

Author: There is no optimal, there is optional. This means that setting this value is not necessary, but will protect against extra slippage if it is set. The minimum is used because it literally means minimum, the swap can result in more. I removed the `desired`.

    2. Requires
        1. An exact amount of either asset to sell
        2. An optional minimum desired output amount
            > **NOTE** This is a safety mechanism against excessive slippage. The [`preview_swap_with_exact_input()`](#preview_swap_with_exact_input) function can be used to calculate a reasonable minimum output amount.
        3. A deadline, i.e., block height limit

#### `swap_with_exact_output()`
1. Allows selling an asset for an exact amount of the other asset 
    1. If
        1. The asset pair of the pool is set
        2. The deadline has not passed
        3. The input amount is sufficient
        4. The asset reserve of asset to buy in the contract is sufficient
        5. The bought amount is not less than the exact output amount
    2. Requires
        1. An exact amount of either asset to buy
        2. A maximum input amount 
            > **NOTE** This is a safety mechanism against excessive slippage. The [`preview_swap_with_exact_output()`](#preview_swap_with_exact_output) function can be used to calculate a reasonable maximum input amount.    
        3. A deadline, i.e., block height limit

> ### Exchange Previews

Reviewer: Similar comments for specificity

Author: Do you mean "other asset" specificity or "minimum / maximum / input / output / optional" specificity here?

Reviewer: Generally anything that can be specific to remove any inferences the better
